File: python/day2-tokens-nft.py
"""
XRPL Academy - Day 2: Tokens (IOUs) and NFTs (Python)
======================================================

WHAT YOU WILL LEARN:
- How to configure an account as a token issuer (Default Ripple)
- How to create a trust line (permission to hold a token)
- How to issue custom tokens
- How to mint an NFT with royalties

BACKGROUND - TOKENS ON XRPL:
Custom tokens on XRPL work differently from XRP (the native currency):
  1. An "issuer" account creates tokens by sending them via a Payment
  2. A "holder" must first create a "trust line" to the issuer
  3. The trust line declares: "I trust this issuer for up to X of this token"
  4. Without a trust line, the payment will fail

This trust-based model prevents token spam -- nobody can force tokens
into your account without your explicit permission.

BACKGROUND - NFTs ON XRPL:
XRPL has built-in NFT support (no smart contracts needed):
  - NFTokenMint creates a new NFT on the ledger
  - NFTs can have royalties (TransferFee) paid on every resale
  - NFTs can be transferable or non-transferable (controlled by Flags)
  - Metadata is stored off-chain, referenced by a URI

PREREQUISITES:
- Install dependencies: pip install xrpl-py
- Run "node scripts/mint-wallets.js" first to create wallets.json
- Complete Day 1 to understand basic transaction flow

RUN: python day2-tokens-nft.py
"""

import json
from xrpl.clients import WebsocketClient
from xrpl.wallet import Wallet
from xrpl.models.transactions import AccountSet, TrustSet, Payment, NFTokenMint
from xrpl.models.amounts import IssuedCurrencyAmount
from xrpl.models.requests import AccountNFTs
from xrpl.transaction import submit_and_wait
from xrpl.utils import str_to_hex

TESTNET_URL = "wss://s.altnet.rippletest.net:51233"

# ---------------------------------------------------------------------------
# STEP 1: Load wallets
# ---------------------------------------------------------------------------
# Wallet 1 = issuer (creates the token and mints the NFT)
# Wallet 2 = holder (receives the token)

with open("../wallets.json") as f:
    data = json.load(f)

issuer = Wallet.from_seed(data["wallets"][0]["seed"])
holder = Wallet.from_seed(data["wallets"][1]["seed"])

print("XRPL Academy - Day 2: Tokens and NFTs (Python)")
print("===============================================")
print("")
print(f"Issuer address: {issuer.address}")
print(f"Holder address: {holder.address}")

with WebsocketClient(TESTNET_URL) as client:
    print(f"Connected to: {TESTNET_URL}")

    # =======================================================================
    # PART 1: CREATE A CUSTOM TOKEN (IOU)
    # =======================================================================
    print("")
    print("--- PART 1: Creating a Custom Token ---")
    print("")

    # -----------------------------------------------------------------------
    # STEP 2: Enable Default Ripple on the issuer account
    # -----------------------------------------------------------------------
    # "Rippling" allows tokens to flow through accounts.
    # An issuer MUST enable Default Ripple so holders can transfer the token
    # to each other (not just back to the issuer).
    #
    # AccountSet with set_flag=8 enables asfDefaultRipple.
    # Flag value 8 = asfDefaultRipple in the XRPL protocol.

    print("Step 1: Enabling Default Ripple on issuer account...")
    print("  (Allows the token to be transferred between holders)")

    account_set_tx = AccountSet(
        account=issuer.address,
        set_flag=8,  # asfDefaultRipple
    )
    result = submit_and_wait(account_set_tx, client, issuer)
    print(f"  Result: {result.result['meta']['TransactionResult']}")

    # -----------------------------------------------------------------------
    # STEP 3: Create a Trust Line from holder to issuer
    # -----------------------------------------------------------------------
    # The holder declares: "I trust the issuer for up to 1,000,000 ACD tokens."
    # This is required before the issuer can send tokens to the holder.
    #
    # TrustSet requires a LimitAmount which specifies:
    #   - currency: 3-character code (e.g., "ACD")
    #   - issuer: the token issuer's address
    #   - value: maximum amount the holder will accept
    #
    # Creating a trust line locks 2 XRP as an "owner reserve" (returned if deleted).

    currency_code = "ACD"  # Change this to your own 3-letter token name

    print("")
    print(f"Step 2: Creating trust line (holder trusts issuer for {currency_code})...")
    print(f"  Limit: 1,000,000 {currency_code}")

    trust_set_tx = TrustSet(
        account=holder.address,
        limit_amount=IssuedCurrencyAmount(
            currency=currency_code,
            issuer=issuer.address,
            value="1000000",
        ),
    )
    result = submit_and_wait(trust_set_tx, client, holder)
    print(f"  Result: {result.result['meta']['TransactionResult']}")

    # -----------------------------------------------------------------------
    # STEP 4: Issue tokens (issuer sends tokens to holder)
    # -----------------------------------------------------------------------
    # Token issuance is simply a Payment from issuer to holder.
    # The issuer "creates" tokens by sending them. There is no separate mint step.
    #
    # For token payments, the amount is an IssuedCurrencyAmount object
    # (not a string of drops like XRP payments).

    print("")
    print(f"Step 3: Issuing 500 {currency_code} tokens to holder...")

    issue_tx = Payment(
        account=issuer.address,
        destination=holder.address,
        amount=IssuedCurrencyAmount(
            currency=currency_code,
            issuer=issuer.address,
            value="500",
        ),
    )
    result = submit_and_wait(issue_tx, client, issuer)
    print(f"  Result: {result.result['meta']['TransactionResult']}")
    print(f"  Holder now has 500 {currency_code}")
    print(f"  View: https://testnet.xrpl.org/accounts/{holder.address}")

    # =======================================================================
    # PART 2: MINT AN NFT
    # =======================================================================
    print("")
    print("--- PART 2: Minting an NFT ---")
    print("")

    # -----------------------------------------------------------------------
    # STEP 5: Mint an NFT using NFTokenMint
    # -----------------------------------------------------------------------
    # NFTokenMint fields:
    #   - nftoken_taxon: numeric category ID (0 for default/testing)
    #   - flags: controls behavior
    #       8 = tfTransferable (NFT can be sold/transferred)
    #   - transfer_fee: royalty in basis points (5000 = 5%)
    #       Creator earns this on every secondary sale
    #   - uri: hex-encoded link to metadata (image, description, etc.)
    #
    # str_to_hex() converts a regular string to hexadecimal encoding,
    # which is how XRPL stores URI data.

    metadata_uri = "https://xrpl.org/img/logo.svg"

    print("Step 4: Minting NFT...")
    print(f"  Minter     : {issuer.address}")
    print(f"  Taxon      : 0 (default collection)")
    print(f"  Transferable: yes")
    print(f"  Royalty    : 5% on every resale")
    print(f"  URI        : {metadata_uri}")

    nft_mint_tx = NFTokenMint(
        account=issuer.address,
        nftoken_taxon=0,
        flags=8,  # tfTransferable
        transfer_fee=5000,  # 5% royalty (basis points)
        uri=str_to_hex(metadata_uri),
    )
    result = submit_and_wait(nft_mint_tx, client, issuer)

    print("")
    print(f"  Result: {result.result['meta']['TransactionResult']}")
    print(f"  View NFTs: https://testnet.xrpl.org/accounts/{issuer.address}/nfts")

print("")
print("Day 2 complete.")
print("You have created a custom token and minted an NFT on the XRPL Testnet.")

File: js/day2-tokens-nft.js
/**
 * XRPL Academy - Day 2: Tokens (IOUs) and NFTs
 * ==============================================
 *
 * WHAT YOU WILL LEARN:
 * - How custom tokens (IOUs) work on XRPL
 * - What "Default Ripple" means and why issuers need it
 * - How trust lines work (permission to hold a token)
 * - How to issue and transfer custom tokens
 * - How to mint an NFT with transfer fees (royalties)
 *
 * BACKGROUND - TOKENS ON XRPL:
 * Unlike XRP (the native currency), custom tokens require a trust relationship:
 *   1. An "issuer" account creates the token by sending it
 *   2. A "holder" account must first create a "trust line" to the issuer
 *   3. The trust line says: "I trust this issuer for up to X units of this token"
 *   4. Only then can the issuer send tokens to the holder
 *
 * This prevents spam -- no one can force unwanted tokens into your account.
 *
 * Token names (currency codes) are 3 characters for standard codes (like "USD",
 * "EUR", "ACD") or up to 20 characters in hex for longer names.
 *
 * BACKGROUND - NFTs ON XRPL:
 * XRPL has native NFT support built into the protocol (no smart contracts needed).
 * NFTs are minted with NFTokenMint and can include:
 *   - A URI pointing to off-chain metadata (image, description, etc.)
 *   - A transfer fee (royalty) paid to the original creator on every resale
 *   - Flags controlling transferability and burnability
 *
 * PREREQUISITES:
 * - Run "node scripts/mint-wallets.js" first to create wallets.json
 * - Complete Day 1 to understand basic transaction flow
 *
 * RUN: node day2-tokens-nft.js
 */

const xrpl = require("xrpl");
const fs = require("fs");

const TESTNET_URL = "wss://s.altnet.rippletest.net:51233";

async function main() {
  // Load wallets -- Wallet 1 will be the token issuer, Wallet 2 the holder
  const { wallets } = JSON.parse(fs.readFileSync("../wallets.json", "utf-8"));
  const issuerWallet = xrpl.Wallet.fromSeed(wallets[0].seed);
  const holderWallet = xrpl.Wallet.fromSeed(wallets[1].seed);

  const client = new xrpl.Client(TESTNET_URL);
  await client.connect();

  console.log("XRPL Academy - Day 2: Tokens and NFTs");
  console.log("======================================");
  console.log("");
  console.log("Issuer address: " + issuerWallet.address);
  console.log("Holder address: " + holderWallet.address);
  console.log("Connected to: " + TESTNET_URL);

  // =========================================================================
  // PART 1: CREATE A CUSTOM TOKEN (IOU)
  // =========================================================================
  console.log("");
  console.log("--- PART 1: Creating a Custom Token ---");
  console.log("");

  // -------------------------------------------------------------------------
  // STEP 1: Enable "Default Ripple" on the issuer account
  // -------------------------------------------------------------------------
  // "Rippling" allows tokens to flow through an account to reach others.
  // An issuer MUST enable Default Ripple so that holders can send the token
  // to each other (not just back to the issuer).
  //
  // Without this flag, tokens would be "stuck" -- holders could only send
  // them back to the issuer, not to other holders.
  //
  // AccountSet is a transaction type that modifies account settings.
  // SetFlag specifies which flag to enable.

  console.log("Step 1: Enabling Default Ripple on issuer account...");
  console.log("  (This allows the token to flow freely between holders)");

  const accountSetResult = await client.submitAndWait(
    {
      TransactionType: "AccountSet",
      Account: issuerWallet.address,
      SetFlag: xrpl.AccountSetAsfFlags.asfDefaultRipple,
    },
    { wallet: issuerWallet }
  );
  console.log("  Result: " + accountSetResult.result.meta.TransactionResult);

  // -------------------------------------------------------------------------
  // STEP 2: Create a Trust Line from the holder to the issuer
  // -------------------------------------------------------------------------
  // A trust line is the holder saying: "I agree to hold up to 1,000,000 units
  // of this token issued by this specific account."
  //
  // The TrustSet transaction creates this relationship.
  // LimitAmount specifies:
  //   - currency: the 3-character token code (you can change "ACD" to anything)
  //   - issuer: who issues this token
  //   - value: maximum amount the holder is willing to hold
  //
  // IMPORTANT: The trust line costs 2 XRP in reserve (locked, not spent).
  // This reserve is returned if the trust line is later deleted.

  const currencyCode = "ACD"; // Change this to your own 3-letter token name

  console.log("");
  console.log("Step 2: Creating trust line from holder to issuer...");
  console.log("  Token: " + currencyCode);
  console.log("  Limit: 1,000,000 " + currencyCode);
  console.log("  (Holder is agreeing to accept this token from the issuer)");

  const trustSetResult = await client.submitAndWait(
    {
      TransactionType: "TrustSet",
      Account: holderWallet.address,
      LimitAmount: {
        currency: currencyCode,
        issuer: issuerWallet.address,
        value: "1000000",
      },
    },
    { wallet: holderWallet }
  );
  console.log("  Result: " + trustSetResult.result.meta.TransactionResult);

  // -------------------------------------------------------------------------
  // STEP 3: Issue tokens (issuer sends tokens to holder)
  // -------------------------------------------------------------------------
  // Token issuance on XRPL is simply a Payment from the issuer to a holder.
  // The issuer "creates" tokens by sending them -- there is no separate
  // minting step. The issuer's balance goes negative (representing obligation)
  // and the holder's balance goes positive.
  //
  // The Amount field for token payments uses an object (not a string like XRP):
  //   { currency: "ACD", issuer: "rIssuerAddress...", value: "500" }

  console.log("");
  console.log("Step 3: Issuing 500 " + currencyCode + " tokens to holder...");
  console.log("  (Issuer creates tokens by sending a Payment)");

  const issueResult = await client.submitAndWait(
    {
      TransactionType: "Payment",
      Account: issuerWallet.address,
      Destination: holderWallet.address,
      Amount: {
        currency: currencyCode,
        issuer: issuerWallet.address,
        value: "500",
      },
    },
    { wallet: issuerWallet }
  );
  console.log("  Result: " + issueResult.result.meta.TransactionResult);
  console.log("  Holder now has 500 " + currencyCode);
  console.log("  View: https://testnet.xrpl.org/accounts/" + holderWallet.address);

  // =========================================================================
  // PART 2: MINT AN NFT
  // =========================================================================
  console.log("");
  console.log("--- PART 2: Minting an NFT ---");
  console.log("");

  // -------------------------------------------------------------------------
  // STEP 4: Mint an NFT using NFTokenMint
  // -------------------------------------------------------------------------
  // NFTokenMint creates a new non-fungible token on the ledger.
  //
  // Key fields:
  //   - NFTokenTaxon: a numeric category/collection ID (0 is fine for testing)
  //   - Flags: controls NFT behavior
  //       8 = tfTransferable (the NFT can be sold/transferred to others)
  //   - TransferFee: royalty percentage in basis points (5000 = 5%)
  //       The creator earns this fee on every secondary sale
  //   - URI: a hex-encoded link to the NFT's metadata (image, description, etc.)
  //       We use convertStringToHex() to encode a URL
  //
  // The NFT is stored directly on the ledger (not in a smart contract).
  // It belongs to the minting account until transferred or burned.

  const metadataUri = "https://xrpl.org/img/logo.svg";

  console.log("Step 4: Minting NFT...");
  console.log("  Minter     : " + issuerWallet.address);
  console.log("  Taxon      : 0 (default collection)");
  console.log("  Transferable: yes");
  console.log("  Royalty    : 5% on every resale");
  console.log("  URI        : " + metadataUri);

  const nftMintResult = await client.submitAndWait(
    {
      TransactionType: "NFTokenMint",
      Account: issuerWallet.address,
      NFTokenTaxon: 0,
      Flags: 8, // tfTransferable -- allows the NFT to be sold/transferred
      TransferFee: 5000, // 5% royalty (in basis points: 5000/100000 = 5%)
      URI: xrpl.convertStringToHex(metadataUri),
    },
    { wallet: issuerWallet }
  );

  console.log("");
  console.log("  Result: " + nftMintResult.result.meta.TransactionResult);
  console.log("  View NFTs: https://testnet.xrpl.org/accounts/" + issuerWallet.address + "/nfts");

  // -------------------------------------------------------------------------
  // DONE
  // -------------------------------------------------------------------------
  await client.disconnect();
  console.log("");
  console.log("Day 2 complete.");
  console.log("You have created a custom token and minted an NFT on the XRPL Testnet.");
}

main().catch(console.error);
